rustc_codegen_ssa: Better code generation for niche discriminants.

mikebenfield opened the pull request:
In some cases we can avoid arithmetic before checking whether a niche is a tag.

Also rename some identifiers around niches.

This is relevant to #101872

rust-highfive commented:
r? @nagisa

(rust-highfive has picked a reviewer for you, use r? to override)

mikebenfield commented:
In my performance testing this improves runtime of every benchmark I tried from `rustc-perf`. Usually around 1% or 2%, sometimes as much as 5%. However I often see discrepancies between the `perf` runs here and my local runs, so who knows. 

I felt some of the names around niches were confusing so I renamed some things. If it's too obnoxious of me to include the renames here I can change them back.

RalfJung commented on compiler/rustc_const_eval/src/interpret/operand.rs:
```diff
@@ -650,7 +650,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         // We use "tag" to refer to how the discriminant is encoded in memory, which can be either
         // straight-forward (`TagEncoding::Direct`) or with a niche (`TagEncoding::Niche`).
-        let (tag_scalar_layout, tag_encoding, tag_field) = match op.layout.variants {
+        let (tag_scalar_layout, encoding, field) = match op.layout.variants {
```
Why did you rename these? The names were carefully chosen. Is 'tag' no longer accurate?

RalfJung commented on compiler/rustc_const_eval/src/interpret/operand.rs:
```diff
@@ -666,8 +666,8 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
-            Variants::Multiple { tag, ref tag_encoding, tag_field, .. } => {
-                (tag, tag_encoding, tag_field)
+            Variants::Multiple { scalar, ref encoding, field, .. } => {
+                (scalar, encoding, field)
```
'scalar', in the interpreter, usually refers to an `interpret::Scalar`. Please avoid using that name for things that have another type.

mikebenfield replied:
Alright. But given that objects of this type shouldn't be referred to as `scalar`s... maybe `Scalar` is not a good name for the type?

RalfJung replied:
I consider its full name to be `abi::Scalar`.

My understanding is that the name makes more sense outside of the interpreter, but given that the interpreter has `interpret::Scalar` we have a somewhat unfortunate conflict.

RalfJung commented on compiler/rustc_const_eval/src/interpret/validity.rs:
```diff
@@ -220,8 +220,8 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
-            Variants::Multiple { tag_field, .. } => {
-                if tag_field == field {
+            Variants::Multiple { field, .. } => {
+                if field == field {
```
Something went very wrong here.

RalfJung commented on compiler/rustc_target/src/abi/mod.rs:
```diff
@@ -1121,38 +1121,44 @@ pub enum Variants<'a> {
-        tag_field: usize,
+        encoding: TagEncoding,
+        scalar: Scalar,
+        field: usize,
```
I could live with removing the `tag_` prefix (but it seems you did a global search-and-replace here and that changed the meaning of some code, see my previous comment), but `scalar: Scalar` is worse than before IMO. This is the scalar ABI of the tag.

You also removed the term 'tag' from the comment, which is bad IMO. We used to have a lot of confusion between discriminant and tag, until I started systematically putting these terms everywhere. Please don't revert us back to the confusing state.

bjorn3 commented on compiler/rustc_codegen_ssa/src/mir/place.rs:
```diff
@@ -222,88 +224,121 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
+                let hi_cmp = bx.icmp(IntPredicate::IntULE, value, hi);
+                bx.assume(lo_cmp);
+                bx.assume(hi_cmp);
+                value
```
Did you forget to apply most of this extra logic to cg_clif?

mikebenfield replied:
I guess cf_clif refers to cranelift. Yeah I made no changes to code generation for cranelift.

RalfJung commented:
I can't comment on the codegen change, but the rename overall is a net negative IMO. I am open to suggestions for how to improve our current names, but we need a term for 'the discriminant in its encoded form in memory'. That thing used to be called 'tag', and this PR leaves that term in some places (`TagEncoding`) but removes it in many others (in particular in many of the comments) without even introducing an alternative term. If you want to replace 'tag' by something else, fine, but we need *some* way to refer to this, and this PR seems to have the goal of just eliminating the term without a replacement and without eliminating the underlying concept.

RalfJung commented:
As I said I am open to renaming 'tag' if you have some good ideas. But I don't like just removing the name without replacement -- concepts like that need names. I'm not saying our current terms are great, they are not, but even bad terms are better than no terms.

The way I think about it: "niche" values of a type are bit patterns that are not valid representations for that type, and can be used to store the tag which can then be mapped to the discriminant. (In that sense they are not even values...)

Commit:
rustc_codegen_ssa: Better code generation for niche discriminants.

In some cases we can avoid arithmetic before checking whether a niche
represents an untagged variant.

This is relevant to #101872

krdln commented:
@mikebenfield Looks like we've both noticed the "needless sub before cmp" and implemented basically the same idea, but you pushed your PR first :laughing:. I actually abandoned my branch, as perf wins I was seeing were very minuscule (despite asm looking quite nice in some artificial tests (see [gist])). I've ran your branch through this gist (see `-mike` file) and interestingly, it regresses in some cases (the `bar` / `baz` etc. functions, which are basically `matches!` on a single variant). The "full-match" (or calling `std::mem::get_discriminant`) is a very nice win though (I think I have a slight regression there)! Anyway, decided to undust and [push my code as a draft](https://github.com/rust-lang/rust/pull/102901).

So, on my branch there are some nice assembly improvements, but almost negligible perf wins, while you have some assembly regressions (and couple improvements) but significant perf wins. This is interesting! Perhaps the case in which I have slight regression, and you have an improvement – a "full-match on u8-based enum", is dominating? I hope we can analyze it further and come up with "best-of-both-worlds" approach.

[gist]: https://gist.github.com/krdln/cefbb4a14d7ddfb5014099d1a0060916

mikebenfield commented:
Oh that's funny we both had the same idea. I didn't check on cases like your `bar` and `baz`; I'm surprised my code regresses in that case. Are you sure your `assume` is always true? `niche_variants` doesn't always have a hole.

krdln commented:
@mikebenfield 
> Are you sure your assume is always true? niche_variants doesn't always have a hole.

You're right. I did a bit of mental shortcut – what I've meant is that `niche_variants` won't ever contain `tagged_variant` (if `tagged_variant` is outside `niche_variants`, this assumption while less useful, will still be valid. I even tried to gate the assume to "true holes only", but it's not worth it).

mikebenfield commented:
OK, here appears to be the reason my codegen regresses in that case. 

Given this LLVM IR:
```LLVM
define i1 @myfunction(i8 %0) {
start:
  %1 = icmp ule i8 %0, 49
  %2 = zext i8 %0 to i64
  %3 = add i64 %2, -46
  %_2 = select i1 %1, i64 %3, i64 0
  %4 = icmp eq i64 %_2, 1
  ret i1 %4
}

```
(notice that `%z` basically is `%0`)
I run an `InstCombinePass` on it: (`opt --instcombine -S nope.ll`)
```LLVM
define i1 @myfunction(i8 %0) {
start:
  %1 = icmp eq i8 %0, 47
  ret i1 %1
}
```

Great! That works as expected. But instead change the function to take an `i16`, and truncate it:
```LLVM
define i1 @myfunction(i16 %0) {
start:
  %z = trunc i16 %0 to i8
  %1 = icmp ule i8 %z, 49
  %2 = zext i8 %z to i64
  %3 = add i64 %2, -46
  %_2 = select i1 %1, i64 %3, i64 0
  %4 = icmp eq i64 %_2, 1
  ret i1 %4
}
```

and the `InstCombinePass` produces this:

```LLVM
define i1 @myfunction(i16 %0) {
start:
  %z = trunc i16 %0 to i8
  %1 = icmp ult i8 %z, 50
  %z.mask = and i16 %0, 255
  %2 = icmp eq i16 %z.mask, 47
  %3 = and i1 %1, %2
  ret i1 %3
}
```

Which is... not so great.

mikebenfield commented:
I filed [this issue](https://github.com/llvm/llvm-project/issues/58275).

mikebenfield commented:
> a "full-match on u8-based enum" – is dominating?

Yeah most of my testing was on matches that are full or nearly full, generally with enough cases so that a jump table is generated.

mikebenfield commented:
Cleaned up `codegen_get_discr` a little, wrote better comments, and made it more careful about when it can move all the arithmetic to after the cast.

Add test `src/test/ui/enum-discriminant/get_discr.rs` to make sure some of those cases work correctly.

nagisa commented:
Before I take a look at this, quantifying the wins is probably a good way to get an objective answer to performance discussions above.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit fa51934b8591ab617b9c03f24c217e815fb64b90 with merge 4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fa51934b8591ab617b9c03f24c217e815fb64b90","merge_sha":"4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8"} -->

mikebenfield commented:
Cool, although note that currently [the LLVM issue I filed](https://github.com/llvm/llvm-project/issues/58275) is probably having a substantial negative impact.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3273114335/jobs/5387452142)
Build commit: 4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8 (`4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3273114335/jobs/5387452142"},"merge_sha":"4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8"} -->

rust-timer commented:
Queued 4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8 with parent e0f8e60dddfecfc9093ee9d9f42557d8260c0355, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e0f8e60dddfecfc9093ee9d9f42557d8260c0355&end=4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8).

rust-timer commented:
Finished benchmarking commit (4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8): [comparison URL](https://perf.rust-lang.org/compare.html?start=e0f8e60dddfecfc9093ee9d9f42557d8260c0355&end=4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=e0f8e60dddfecfc9093ee9d9f42557d8260c0355&end=4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 0.3%     | [0.2%, 0.6%] | 41        |
| Regressions ❌ <br /> (secondary) | 0.5%     | [0.2%, 1.7%] | 23        |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 0.3%     | [0.2%, 0.6%] | 41        |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e0f8e60dddfecfc9093ee9d9f42557d8260c0355&end=4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | -        | -     | 0         |
| Regressions ❌ <br /> (secondary) | -        | -     | 0         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -3.1%    | [-5.1%, -2.1%] | 25        |
| All ❌✅ (primary) | -        | -     | 0         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e0f8e60dddfecfc9093ee9d9f42557d8260c0355&end=4b2f5a48d62e8bf41962efc9cc6efffe3a994fe8&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

krdln commented:
@mikebenfield Looks like you still do have regressions on the `is_x()`-style functions. Perhaps this is the contributor to the regressions? Check out this gist https://gist.github.com/krdln/262f932d8ae75f19673bae3b2165f1c1.

Another hypothesis – perhaps too much of assumes do hurt? https://llvm.org/docs/Frontend/PerformanceTips.html

> Avoid using the [assume intrinsic](https://llvm.org/docs/LangRef.html#int-assume) until you’ve established that a) there’s no other way to express the given fact and b) that fact is critical for optimization purposes. Assumes are a great prototyping mechanism, but they can have negative effects on both compile time and optimization effectiveness. The former is fixable with enough effort, but the later is fairly fundamental to their designed purpose.

I had a case of `assume` pessimizing codegen (will try to create an issue later) on my branch.

mikebenfield commented:
Interesting. It's certainly the case that LLVM isn't making good use of the `assume`, so maybe I should just get rid of it.

nagisa commented:
The assumes should definitely be added sparingly as they can pessimize other passes that don't handle it or are forced to preserve it in the patterns which would otherwise end up getting removed outright.

Unless there is a demonstrable improvement in IR quality at various stages of optimization, they should not be output at all.

I’d suggest adding a comment to the assume, documenting those improvements, so that there is a clear pathway to removing them (anybody can check the pattern to see if the assumes are still doing anything).

mikebenfield commented:
Here's some benchmark data. I benchmarked 3 compilers: `master`, `bgd-assume` (this branch cherry picked onto `master`), and `bgd-noassume` (same but with the `assume` lines commented out).

[Here](https://gist.github.com/mikebenfield/3204ab7ef57f556b4aefb261cdf8f772) is some code for a microbenchmark, which does many matches on a simple enum. 

On my machine (running this like `cargo +master release -- 10000000`), I consistently get around

96ms for master

90ms for bgd-assume

90ms for bgd-noassume

As for benchmarking the compiler itself using the `rustc-perf` tool, here are some instruction counts:
```
('issue-46449', 'Opt-IncrFull')
1,014,347,557 (master)
1,031,204,445 (assume)
988,152,707 (noassume)

('regex', 'Opt-IncrPatched0')
38,185,135,497 (master)
38,336,547,819 (assume)
38,312,812,860 (noassume)

('regex', 'Opt-IncrFull')
88,287,034,172 (master)
88,500,354,715 (assume)
88,457,370,856 (noassume)
```

And here's some CPU time:

```
('issue-46449', 'Opt-IncrFull')
742.517084ms (master)
715.239732ms (assume)
692.373249ms (noassume)

('regex', 'Opt-IncrPatched0')
16.20162392s (master)
15.84068765s (assume)
16.601834155s (noassume)

('regex', 'Opt-IncrFull')
79.921668087s (master)
80.0554681s (assume)
79.500080144s (noassume)
```

I guess based on the instruction counts in particular it does seem best to get rid of the `assume`s. I'll upload a revision with those gone.

nagisa reviewed (commented):
Inline comments are all largely nits. I haven't been able to find anything partiuclarly wrong with the algorithm, but I would like to see a codegen test (`src/tests/codegen`) added in addition to the UI test that would exercise the three special cases.

I would imagine that the test would basically be a sequence of `CHECK`s in basically the same order as in the implementation. Unfortunately we don’t have the same test case check generator that LLVM does :(

nagisa commented:
Lets do another perf run now that assumes are gone. I feel like having _some_ regressions with a change as far reaching as this is probably going to be unavoidable. But having some numbers to look at in T-compiler meeting is still necessary, in case there are regressions.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 623f1636b4c6c16408df46705c83a2228f5fdfaa with merge e2c38dd4de47bf88237fedb5223489e39d70ba44...
<!-- homu: {"type":"TryBuildStarted","head_sha":"623f1636b4c6c16408df46705c83a2228f5fdfaa","merge_sha":"e2c38dd4de47bf88237fedb5223489e39d70ba44"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3397907065/jobs/5651407607)
Build commit: e2c38dd4de47bf88237fedb5223489e39d70ba44 (`e2c38dd4de47bf88237fedb5223489e39d70ba44`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3397907065/jobs/5651407607"},"merge_sha":"e2c38dd4de47bf88237fedb5223489e39d70ba44"} -->

rust-timer commented:
Queued e2c38dd4de47bf88237fedb5223489e39d70ba44 with parent 09508489efc223287731fe8abbd2a81bbf7adf8e, future [comparison URL](https://perf.rust-lang.org/compare.html?start=09508489efc223287731fe8abbd2a81bbf7adf8e&end=e2c38dd4de47bf88237fedb5223489e39d70ba44).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (e2c38dd4de47bf88237fedb5223489e39d70ba44): [comparison URL](https://perf.rust-lang.org/compare.html?start=09508489efc223287731fe8abbd2a81bbf7adf8e&end=e2c38dd4de47bf88237fedb5223489e39d70ba44&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=09508489efc223287731fe8abbd2a81bbf7adf8e&end=e2c38dd4de47bf88237fedb5223489e39d70ba44&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.2%, 0.4%]   | 7     |
| Regressions ❌ <br /> (secondary)  | 2.1%  | [2.1%, 2.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.7% | [-2.2%, -1.4%] | 3     |
| All ❌✅ (primary)                 | 0.3%  | [0.2%, 0.4%]   | 7     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=09508489efc223287731fe8abbd2a81bbf7adf8e&end=e2c38dd4de47bf88237fedb5223489e39d70ba44&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.8%  | [0.7%, 0.8%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 4.7%  | [2.3%, 7.0%]   | 2     |
| Improvements ✅ <br /> (primary)   | -2.2% | [-3.2%, -0.9%] | 3     |
| Improvements ✅ <br /> (secondary) | -1.6% | [-1.6%, -1.6%] | 1     |
| All ❌✅ (primary)                 | -1.0% | [-3.2%, 0.8%]  | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=09508489efc223287731fe8abbd2a81bbf7adf8e&end=e2c38dd4de47bf88237fedb5223489e39d70ba44&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.7%, -0.7%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.0% | [-3.6%, -2.1%] | 4     |
| All ❌✅ (primary)                 | -0.7% | [-0.7%, -0.7%] | 1     |
</details>

nagisa commented:
Looking at the primary benchmarks:

I am failing to find a meaningful difference – the `check` builds regressing might be a second order effect? But actual absolute numbers are so low it is really hard to actually spot any difference.

For `opt` builds the majority of increase is in LTO/optimization passes in LLVM, which I would say is probably expected given that the shape of code has changed slightly and LLVM likely simply has more opportunities to do things instead of giving up. Change in absolute numbers is, again, very small.

The same stands for the big regression in the secondary `issue-46449 | opt | incr-patched: empty 3072` benchmark where the majority of the change seems to be concentrated in LLVM. I believe we can justify this, given this same test in three other scenarios is showing an improvement of a similar degree.

Considering that, and also the improvement shown by microbenchmarks above, I believe that ultimately this is okay to land.

mikebenfield commented on compiler/rustc_codegen_ssa/src/mir/place.rs:
```diff
@@ -242,68 +245,156 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
+                let tag_size = tag_scalar.size(bx.cx());
+                let max_unsigned = tag_size.unsigned_int_max();
+                let max_signed = max_unsigned >> 1;
+                let min_signed = max_signed + 1;
```
I changed to use `signed_int_max`, but `signed_int_min` does the wrong thing: all the high bits in the `i128` are set.

mikebenfield commented on compiler/rustc_codegen_ssa/src/mir/place.rs:
```diff
@@ -242,68 +245,156 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
+                    (lhs ^ min_signed) <= (rhs ^ min_signed)
+                };
+
+                // We have a subrange `niche_start..=niche_end` inside `range`.
```
If so, I think it should probably happen when a niche is created.

mikebenfield commented:
What in particular would you like to see in the `codegen` test?

nagisa commented:
> What in particular would you like to see in the codegen test?

The LLVM-IR for the discriminant reads that exercise the three different cases implemented in this PR, post basic optimization. That way any future changes that would regress (or improve) the codegen here would get properly flagged by the test suite. A regression test of sorts for the referenced issue.

Probably something like this:

```
// compile-flags: -Copt-level=1
// min-llvm-version: 15.0
// ^^^^^^^^^^^^^^^^^^^^^^ if necessary

// CHECK-LABEL: match_case_simple
// CHECK: icmp eq ... or whatever LLVM spits out today
// CHECK-NEXT: ...
// ...
#[no_mangle]
pub fn match_case_simple(enumeration: SimpleEnum) {
    match enumeration { ... } // or std::mem::discriminant()?
}

// ... And similiarly for the other two cases
```

mikebenfield commented:
Alright, I will add such a test soon.

mikebenfield commented:
Added a codegen test. I'm just using the names created by LLVM explicitly; let me know if there's a preferabe way.

nagisa commented:
I think its fine for the time being, even if this isn’t particularly future proof – if it ends up being a pain point, the test can always be adjusted in the future.

@bors r+

bors commented:
:pushpin: Commit 2adb8178ce9b0d6a8a384d5e5ee3f2ca66c6deeb has been approved by `nagisa`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nagisa 2adb8178ce9b0d6a8a384d5e5ee3f2ca66c6deeb -->
<!-- homu: {"type":"Approved","sha":"2adb8178ce9b0d6a8a384d5e5ee3f2ca66c6deeb","approver":"nagisa","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 2adb8178ce9b0d6a8a384d5e5ee3f2ca66c6deeb with merge f25056b45fbe354ffc0e56b0429739de0d6fe396...
<!-- homu: {"type":"BuildStarted","head_sha":"2adb8178ce9b0d6a8a384d5e5ee3f2ca66c6deeb","merge_sha":"f25056b45fbe354ffc0e56b0429739de0d6fe396"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3424709228/jobs/5704741332) [(plain)](https://github.com/rust-lang-ci/rust/commit/f25056b45fbe354ffc0e56b0429739de0d6fe396/checks/9377231988/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

mikebenfield commented:
@nagisa I assume that failure is some network/infrastructure hiccup? Otherwise I'm not sure what's going on.

lqd commented:
yes

@bors retry network failure in aarch64-gnu builder: "curl: (35) OpenSSL SSL_connect: Connection reset by peer in connection to ci-mirrors.rust-lang.org:443"

bors commented:
:hourglass: Testing commit 2adb8178ce9b0d6a8a384d5e5ee3f2ca66c6deeb with merge 661023d5027c336e0aabe15e136c0f38d50146f4...
<!-- homu: {"type":"BuildStarted","head_sha":"2adb8178ce9b0d6a8a384d5e5ee3f2ca66c6deeb","merge_sha":"661023d5027c336e0aabe15e136c0f38d50146f4"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3439827085/jobs/5737830629)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3439827085/jobs/5737830629","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-i586-gnu-i586-i686-musl`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3439827085/jobs/5737567133) [(plain)](https://github.com/rust-lang-ci/rust/commit/661023d5027c336e0aabe15e136c0f38d50146f4/checks/9420843342/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] src/test/codegen/enum-match.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum-match/enum-match.ll" "/checkout/src/test/codegen/enum-match.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC"
stdout: none
--- stderr -------------------------------
/checkout/src/test/codegen/enum-match.rs:37:16: error: CHECK-NEXT: expected string not found in input
// CHECK-NEXT: %2 = zext i8 %0 to i64
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum-match/enum-match.ll:18:24: note: scanning from here
 %1 = icmp ugt i8 %0, 1
                       ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum-match/enum-match.ll:19:2: note: possible intended match here
 %2 = zext i8 %0 to i32

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/enum-match/enum-match.ll
Check file: /checkout/src/test/codegen/enum-match.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
           .
           .
           .
           .
          13: } 
          14:  
          15: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind readnone willreturn uwtable 
          16: define i8 @match1(i8 noundef %0) unnamed_addr #0 { 
          17: start: 
          18:  %1 = icmp ugt i8 %0, 1 
next:37'0                            X error: no match found
          19:  %2 = zext i8 %0 to i32 
next:37'0     ~~~~~~~~~~~~~~~~~~~~~~~~
next:37'1      ?                       possible intended match
          20:  %3 = add nsw i32 %2, -1 
next:37'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
          21:  %_2 = select i1 %1, i32 %3, i32 0 
next:37'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          22:  switch i32 %_2, label %bb2 [ 
next:37'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          23:  i32 0, label %bb3 
next:37'0     ~~~~~~~~~~~~~~~~~~~
          24:  i32 1, label %bb5 
next:37'0     ~~~~~~~~~~~~~~~~~~~
           .
           .
>>>>>>
------------------------------------------
```

</details>

mikebenfield commented:
Alright, that failed because the LLVM IR I test against is not what's generated on 32 bit systems. Duh. I added `// only-x86_64` to the test; someone let me know if that is not an appropriate fix.

nagisa commented:
It is better to have a `#[no_core]` test that specifies `--target=x86_64-unknown-linux-gnu` or somesuch so that the test can run locally on developers’ non-x86-64 machines, but this is okay too.

@bors r+

bors commented:
:pushpin: Commit 51918dcc51360d79cec771ef556ae9181ebaeb2e has been approved by `nagisa`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nagisa 51918dcc51360d79cec771ef556ae9181ebaeb2e -->
<!-- homu: {"type":"Approved","sha":"51918dcc51360d79cec771ef556ae9181ebaeb2e","approver":"nagisa","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 51918dcc51360d79cec771ef556ae9181ebaeb2e with merge 742d3f02c243964e5b868d90afd60c2907be5853...
<!-- homu: {"type":"BuildStarted","head_sha":"51918dcc51360d79cec771ef556ae9181ebaeb2e","merge_sha":"742d3f02c243964e5b868d90afd60c2907be5853"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3445498368/jobs/5751863412)
Approved by: nagisa
Pushing 742d3f02c243964e5b868d90afd60c2907be5853 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3445498368/jobs/5751863412"},"merge_sha":"742d3f02c243964e5b868d90afd60c2907be5853"} -->

rust-timer commented:
Finished benchmarking commit (742d3f02c243964e5b868d90afd60c2907be5853): [comparison URL](https://perf.rust-lang.org/compare.html?start=b7b7f2716ee1655a696d3d64c3e12638d0dd19c0&end=742d3f02c243964e5b868d90afd60c2907be5853&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=b7b7f2716ee1655a696d3d64c3e12638d0dd19c0&end=742d3f02c243964e5b868d90afd60c2907be5853&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.5%, 0.6%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.1%  | [2.1%, 2.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-2.2%, -0.3%] | 4     |
| All ❌✅ (primary)                 | 0.5%  | [0.5%, 0.6%]   | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b7b7f2716ee1655a696d3d64c3e12638d0dd19c0&end=742d3f02c243964e5b868d90afd60c2907be5853&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=b7b7f2716ee1655a696d3d64c3e12638d0dd19c0&end=742d3f02c243964e5b868d90afd60c2907be5853&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.9%, 1.9%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.8% | [-1.8%, -1.7%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

nnethercote commented:
This must have been right on the edge of the regression/no-regression categorization. Not much to worry about here.

@rustbot label: +perf-regression-triaged
